### ISSUE TITLE
`users.users.<name>.shell` does not update the user's shell

### ISSUE THREAD

**Issue by Samuel-Martineau** [LnL7/nix-darwin#811]
For example, setting this in `~/.config/nix-darwin/flake.nix` still requires running `chsh -s /run/current-system/sw/bin/fish`.
```nix
programs.fish.enable = true;

environment.shells = [pkgs.fish];

users.users.${mainUser} = {
  home = "/Users/${mainUser}";
  shell = pkgs.fish;
};
```
Also, should setting `users.users.<name>.shell` automatically add the shell to `environment.shells`?

**Comment by Samuel-Martineau** [comment]
Closing as duplicate of #122

**Comment by Enzime** [comment]
This looks to me like a separate to the one in #122

**Comment by bestlem** [comment]
@Enzime I agree that this should be separate - the issue I think is that [one](https://github.com/LnL7/nix-darwin/issues/122#issuecomment-1782971499) (out of the many ) comments there did put the two issues together and this is now seen - I think that it would be easier to see the issues to ignore that comment.

**Comment by Samuel-Martineau** [comment]
@bestlem I am not even sure they are related anymore: even after manually changing my shell on my machine, the path issue was not fixed.

**Comment by thuvasooriya** [comment]
What is the best way to resolve this for now? To set fish as the default login shell using nix?

**Comment by spcutrell** [comment]
Has this been resolved?

**Comment by Enzime** [comment]
This should’ve been fixed by #818

**Comment by spcutrell** [comment]
> This should’ve been fixed by #818

So now, in theory, I would just need:

    programs.fish.enable = true;
    users.users = {
      <name> = {
        home = "/Users/<name>";
        shell = pkgs.fish;
      };
    };

and it should "just work"?

**Comment by Enzime** [comment]
Yeah that should work, feel free to comment if it doesn’t work

**Comment by spcutrell** [comment]
It definitely doesn't work for me. I see the correct location in `/etc/shells` at `/run/current-systems/sw/bin/fish`. There is a very real chance it is a GIGO issue. I pushed my proof-of-concept config if that is helpful to rule out User error.

https://github.com/spcutrell/nix/tree/main

**Comment by spcutrell** [comment]
Hoping to reopen this issue. It doesn't seem fixed for me.

Steps to recreate:

1. Fresh install on M1 (Sonoma 14.5)
2. Nix installation using Determinate Installer
3. Create `~/.config/nix/flake.nix` (see below for content)
4. Create `~/.config/nix/modules/darwin/default.nix` (see below for content)
5. Run `nix run nix-darwin -- switch --flake ~/.config/nix/`
6. Run `run/current-system/sw/bin/darwin-rebuild switch --flake ~/.config/nix` for good measure
7. Reboot for good measure

```
{
  # ~/.config/nix/flake.nix
  description = "Entry flake";

  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";

    darwin.url = "github:LnL7/nix-darwin";
    darwin.inputs.nixpkgs.follows = "nixpkgs";
  };

  outputs = inputs@{ self, nixpkgs, darwin, home-manager, ... }: {
    darwinConfigurations."My-MacBook-Air" = darwin.lib.darwinSystem {
      system = "aarch64-darwin";
      pkgs = import nixpkgs {
        system = "aarch64-darwin";
        config.allowUnfree = true;
      };
      modules = [ ./modules/darwin ];
    };
  };
}
```

```
{ pkgs, ... } : {
  # ~/.config/nix/modules/darwin/default.nix
  programs.fish.enable = true;
  
  environment.shells = [pkgs.fish];

  users.users.myUser = {
    home = /Users/myUser;
    shell = pkgs.fish;
  };
  nix.extraOptions = ''
    experimental-features = nix-command flakes
  '';
  services.nix-daemon.enable = true;
  system.stateVersion = 4;
}
```

`cat /etc/shells` to prove that the build is working: 
```# List of acceptable shells for chpass(1).
# Ftpd will not allow users to connect who are not using
# one of these shells.

/bin/bash
/bin/csh
/bin/dash
/bin/ksh
/bin/sh
/bin/tcsh
/bin/zsh

# List of shells managed by nix.
/run/current-system/sw/bin/fish
```

Now from here, I am not really sure what I expect. However, opening up terminal is still zsh and `echo $SHELL` still shows `/bin/zsh` and `chsh -s /run/current-system/sw/bin/fish` correctly sets shell to fish as expected (showing that nix properly installed it).

**Comment by Enzime** [comment]
I don't have the time to investigate and fix this now, I'll just leave some notes here for anyone who wants to test it out and make a PR

I suspect it'll have something to do with this command

https://github.com/LnL7/nix-darwin/blob/cf297a8d248db6a455b60133f6c0029c04ebe50e/modules/users/default.nix#L175

Possibly `-create` doesn't work if the user already exists/shell is already defined and we might need to use a different flag instead for `dscl`

**Comment by z0al** [comment]
> Possibly -create doesn't work if shell is already defined and we might need to use a different flag instead for dscl

@Enzime when I run the same command manually it works fine even if the shell is already set
```
sudo dscl . -create /Users/me UserShell /run/current-system/sw/bin/fish
```

I [tried to to debug](https://github.com/z0al/nix-darwin/blob/841366a3035fccb8aa917a696918dffcbf277f4c/modules/users/default.nix#L175-L176) the path being set by adding an `echo` I don't see it logged any where during system activation. What am I missing?

**Comment by Enzime** [comment]
I think the only way that code doesn’t run is if your user’s uid doesn’t match the one set in nix-darwin (I think the default is 501), if that’s the case a warning should be emitted

**Comment by z0al** [comment]
My user ID is 501 🤔 

```
$ dscl . -read /Users/z0al UniqueID
UniqueID: 501
```

**Comment by Enzime** [comment]
Can you test if the `dscl` command can change your shell to `zsh` or `bash`? You might need to restart before the change is reflected

**Comment by z0al** [comment]
Yep, I did that and it works without restarting.

```
$ dscl . -read /Users/z0al UserShell
UserShell: /run/current-system/sw/bin/fish

$ sudo dscl . -create /Users/z0al UserShell /run/current-system/sw/bin/bash

$ dscl . -read /Users/z0al UserShell
UserShell: /run/current-system/sw/bin/bash

$ sudo dscl . -create /Users/z0al UserShell /run/current-system/sw/bin/fish

$ dscl . -read /Users/z0al UserShell
UserShell: /run/current-system/sw/bin/fish
```

**Comment by spcutrell** [comment]
To throw a wrinkle into this, my shell doesn't change using the `dscl` command. UserId is 501.

**Comment by z0al** [comment]
That's weird. In my case, this works just fine. Can you give it a shot @spcutrell ?

```nix
  system.activationScripts.setFishAsShell.text = ''
    dscl . -create /Users/${user} UserShell /run/current-system/sw/bin/fish
  '';
```

**Comment by spcutrell** [comment]
Well... forget what I posted. It is working now. I thought I was pretty thorough, but I'm going to chalk this up to user error for now.

The `dscl . -create` command works.

**Comment by z0al** [comment]
I figured it out (at least for my own setup). As I mentioned earlier, the activation script doesn't run at all since I don't see logs e.g [setting up users](https://github.com/LnL7/nix-darwin/blob/5ce8503cf402cf76b203eba4b7e402bea8e44abc/modules/users/default.nix#L146)

Looking at the script, it's behind a condition `mkIf (cfg.knownUsers != [])` 🤦🏻‍♂️ . So the fix was as simple as adding this to my config:

```nix
users.knownUsers = [ "z0al" ];
users.users.z0al.uid = 501;
```

And boom! I now see the logs and the shell is correctly set up 🎉

**Comment by spcutrell** [comment]
Confirming that it works for me as well.

I'm not sure what mac views as "admin", but is there any concern about this from the docs?

[users.knownUsers](https://daiderd.com/nix-darwin/manual/index.html#opt-users.knownUsers)
List of users owned and managed by nix-darwin. Used to indicate what users are safe to create/delete based on the configuration. _Don’t add the admin user or other system users to this_.

**Comment by z0al** [comment]
> I'm not sure what mac views as "admin", but is there any concern about this from the docs?

I'm also not sure. My user is definitely an admin and I'd expect every other single-user setups to have the same issue.

**Comment by spcutrell** [comment]
Well in any case, great work and thanks for your help. Not sure if this is expected behavior or how the maintainers would like to proceed, but it is pseudo-resolved imo.

Side note: I decided I would learn Nix by just slowly building out a nix-darwin set-up, and my first goal was to get the shell changed from zsh... didn't realize I stepped into a trap!

**Comment by z0al** [comment]
Funnily enough, I also stumbled across the same issue last year while building my nix-darwin/nixos setup. I made a home-manager script that calls `chsh -s /path/to/fish` at the time and it worked just fine. 

Today, however, I was doing a major refactoring to my setup and and decided to check if the issue is still open 😄

**Comment by Samasaur1** [comment]
I wasn't around when this option was created, so I don't know for sure, but my instinct is that that warning is there because letting nix-darwin manage your primary/admin user could potentially really screw up your system. I would strongly recommend **not** putting your user in `users.knownUsers` and would instead run `dscl . -create /Users/${user} UserShell /run/current-system/sw/bin/fish` or `chsh -s /run/current-system/sw/bin/fish` in an activation script

**Comment by z0al** [comment]
It's not just about the shell though, most if not all of the options under `users.users.<user>` are set within that same script so that renders all of them useless?

**Comment by emilazy** [comment]
Yes, I think the status quo is that you shouldn’t use the `users.users.*` arguments on your main user, but frankly I forget why.

**Comment by z0al** [comment]
>  you shouldn’t use the `users.users.*` arguments on your main user

Coming from NixOS background that's quite the surprise frankly. 

I suggest making that more prominent in the official docs to avoid confusion

**Comment by emilazy** [comment]
We could maybe just remove the warning. It seems like the worry was that removing your user config would delete your user, but you can’t actually delete the currently logged‐in user anyway.

**Comment by z0al** [comment]
> It seems like the worry was that removing your user config would delete your user,

I'd expect nix-darwin to _manage_ the user if specified in `users.users.*` and ignore it if it's not there by default. That's at least the default behavior in NixOS AFAIK unless you explicity set [users.mutableUsers](https://nixos.org/manual/nixos/stable/options#opt-users.mutableUsers) to false

**Comment by eblume** [comment]
Hi, I've just run in to some related issue to this, and reading this ticket helped me get to a fix that I want to share with others. I don't fully understand what was going on, but I can guess. My fix was simply:

```console
$ chsh -s /run/current-system/sw/bin/fish
```

The root of my issue, I believe, is that I had previously done something like:

```console
$ brew install fish
$ sudo echo "/opt/homebrew/bin/fish" >> /etc/shells
$ chsh -s /opt/homebrew/bin/fish
```

I'm not sure I fully understand the resulting effects I saw in my `$SHELL`, `$PATH`, and in the way that fish "configured itself", but it seems like one downstream effect of it was removing any instance of  `/run/current-system` from my `$PATH`. Further I've read this bug thread and its linked tickets and I'm not sure I understand the implications of `users.users.${user}.shell` as it seems like this was being specified but not applied without any error.

Please let me know if this information aught to be directed elsewhere. Thanks and good luck!

**Comment by kblcuk** [comment]
Fwiw I had the same problem, and it might be that for me this was somehow caused by homebrew installation of fish. Which, after uninstalling it, didn't clear the `UserShell`:

```
❯ dscl . -read /Users/kblcuk UserShell
UserShell: /opt/homebrew/bin/fish

~/nix-darwin-config on ☁️  (eu-central-1)
❯ sudo dscl . -create /Users/kblcuk UserShell /run/current-system/sw/bin/fish

~/nix-darwin-config on ☁️  (eu-central-1)
❯ dscl . -read /Users/kblcuk UserShell
UserShell: /run/current-system/sw/bin/fish
```

**Comment by Enzime** [comment]
Fixed in #1120

**Comment by alankritjoshi** [comment]
> Fwiw I had the same problem, and it might be that for me this was somehow caused by homebrew installation of fish. Which, after uninstalling it, didn't clear the `UserShell`:
> 
> ```
> ❯ dscl . -read /Users/kblcuk UserShell
> UserShell: /opt/homebrew/bin/fish
> 
> ~/nix-darwin-config on ☁️  (eu-central-1)
> ❯ sudo dscl . -create /Users/kblcuk UserShell /run/current-system/sw/bin/fish
> 
> ~/nix-darwin-config on ☁️  (eu-central-1)
> ❯ dscl . -read /Users/kblcuk UserShell
> UserShell: /run/current-system/sw/bin/fish
> ```

I ran into this exact thing. Had homebrew fish uninstalled. The fix in https://github.com/nix-darwin/nix-darwin/pull/1120 didn't seem to fix this weirdness with fish
